src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComponentActivator.cs

```diff
@@ -0,0 +1,197 @@
+                default:
+                    throw new ArgumentException(nameof(stringFormat));
+            }
+        }
+
+        private static CreateDelegateArgs CheckAndMarshalNativeInput(CreateNativeDelegateArgs argsNative)
+        {
+            if (argsNative.assemblyPath == IntPtr.Zero)
+            {
+                throw new ArgumentNullException(nameof(argsNative.assemblyPath));
+            }
+
+            if (argsNative.typeName == IntPtr.Zero)
+            {
+                throw new ArgumentNullException(nameof(argsNative.typeName));
+            }
+
+            if (argsNative.methodName == IntPtr.Zero)
+            {
+                throw new ArgumentNullException(nameof(argsNative.methodName));
+            }
+
+            if (argsNative.delegateType == IntPtr.Zero)
+            {
+                throw new ArgumentNullException(nameof(argsNative.delegateType));
+            }
+
+            if (argsNative.functionHandle == IntPtr.Zero)
+            {
+                throw new ArgumentNullException(nameof(argsNative.functionHandle));
+            }
+
+            StringFormat stringFormat = (StringFormat) argsNative.stringFormat;
+
+            CreateDelegateArgs args = new CreateDelegateArgs()
+            {
+                assemblyPath = MarshalToString(stringFormat, argsNative.assemblyPath),
+                typeName     = MarshalToString(stringFormat, argsNative.typeName),
+                methodName   = MarshalToString(stringFormat, argsNative.methodName),
+                delegateType = MarshalToString(stringFormat, argsNative.delegateType),
+            };
+
+            return args;
+        }
+
+        /// <summary>
+        /// Native hosting entry point for creating a native delegate
+        /// </summary>
+        /// <param name="argsNative">Reference to a <see cref="CreateNativeDelegateArgs"/> instance</param>
+        [CLSCompliant(false)]
+        public unsafe static int CreateNativeDelegate(CreateNativeDelegateArgs argsNative)
+        {
+            CreateDelegateArgs args;
+
+            try
+            {
+                args = CheckAndMarshalNativeInput(argsNative);
+
+                Delegate d = CreateDelegate(ref args);
+
+                IntPtr functionPtr = Marshal.GetFunctionPointerForDelegate(d);
+
+                lock(s_Delegates)
+                {
+                    s_Delegates[functionPtr] = d;
+                }
+
+                Marshal.WriteIntPtr(argsNative.functionHandle, functionPtr);
+            }
+            catch (Exception e)
+            {
+                return e.HResult;
+            }
+
+            return 0;
+        }
+
+        /// <summary>
+        /// Native hosting entry point for releasing a created native delegate
+        /// </summary>
+        /// <param name="functionPtr">Native delegate to be released</param>
+        [CLSCompliant(false)]
+        public unsafe static int ReleaseNativeDelegate(IntPtr functionPtr)
```

Comment:
> how is this going to be used from the native code

I was expecting the ReleaseDelegate to be exposed in the native hosting API.

> OK to not have a free story

Seems like now or never.

Reply:
I don't understand why "Now or never". We didn't have ALC unloadability for 2 releases and it work reasonably well. Adding it should not be breaking in any way if its opt-in.

That said I'm not against trying to solve this for real and add support for unloadability right away.

Reply:
> I don't understand why "Now or never"

It becomes a public API as soon as we ship it. A different API will need to be added to support unloadability.

Removed it.  Not relevant for this iteration.

Reply:
That was why I was advocating for having a `reserved` parameter, so that we can extend the "public" API after we ship it without breaking anything. It still doesn't solve how to get the "FreeDelegate" function pointer to native, but that's a separate issue.